Add response templates to user settings

Zhao-Andy opened the pull request:
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This adds response templates to the settings page. Note that it does not make a link visible, but it is technically available via the direct route `/settings/response-templates`.

Users who access this page will be able to complete basic CRUD for personal response templates.

Also removes some dead CSS.

## Related Tickets & Documents
#5419 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
![Screen Shot 2020-03-24 at 5 24 35 PM](https://user-images.githubusercontent.com/17884966/77478451-5f81a080-6df4-11ea-94ee-a4556682a9f9.png)


## Added tests?
- [x] yes

## Added to documentation?
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
No

## [optional] What gif best describes this PR or how it makes you feel?

![Kermit plays the banjo with Fozzie, the two in a car ride together with the text "We're close".](https://media.giphy.com/media/VEzlrMWk3F7uuFuRSq/source.gif)

Zhao-Andy commented on app/assets/stylesheets/settings.scss:
```diff
@@ -545,30 +550,29 @@ $input-width: 650px;
+          text-align: center;
+        }
       }
-    }
-    .github-repo-fork {
-      margin-left: 5px;
-      font-weight: 400;
-      background: $purple;
-      color: $bold-blue;
-      border-radius: 3px;
-      padding: 1px 6px;
-      font-size: 0.9em;
-    }
-    .github-repo-featured-indicator {
-      display: inline-block;
-      float: right;
-      background: $green;
-      color: white;
-      border: 0px;
```
This is dead code.

madhavarshney replied:
Are you sure? It seems to be used here:

https://github.com/thepracticaldev/dev.to/blob/81fff248dd953f873161ed276ddb7f0ab576bcf8/app/javascript/githubRepos/singleRepo.jsx#L13-L15

This may have introduced #8834.

Zhao-Andy replied:
Ahh yeah, I think you're right. Good catch!

Zhao-Andy commented on app/controllers/response_templates_controller.rb:
```diff
@@ -0,0 +1,74 @@
+class ResponseTemplatesController < ApplicationController
+  after_action :verify_authorized, except: %i[index]
+
+  def index
```
I'm not that big of a fan of this logic, but since it worked previously I decided to keep it for now. Will spend some time to refactor this in the cycle.

vaidehijoshi commented on app/controllers/response_templates_controller.rb:
```diff
@@ -0,0 +1,74 @@
+  after_action :verify_authorized, except: %i[index]
+
+  def index
+    not_found unless current_user
```
you could pull this out into a method and then `before_action` it

vaidehijoshi commented on app/controllers/response_templates_controller.rb:
```diff
@@ -0,0 +1,74 @@
+                            where(user_id: nil, type_of: "mod_comment").
+                            union(user_id: current_user.id, type_of: "personal_comment")
+                          handle_authorization(result)
+                          result
```
Does the `authorize` in the handle_authorization [method](https://github.com/thepracticaldev/dev.to/pull/6823/files#diff-e6c099ad651ac2c674056e5f8340a483R71) raise an error if it can't be authorized?

This reads a little funny to me; is there any reason to not authorize the `result` and return it from the same method? (Rather than assigning it, passing it to the `handle_authorization` method, and then returning it from the `index` method?

Zhao-Andy replied:
👍 all good thoughts here. Think I will refactor this after all 😓

vaidehijoshi commented on app/controllers/response_templates_controller.rb:
```diff
@@ -0,0 +1,74 @@
+      authorize records, :moderator_index?
+    else
+      authorize records, :admin_index?
+    end
```
Should this be a generic `else` here or an `elsif`? Wondering what happens if someone passes in a `params[:type_of]` that we're not expecting?

Zhao-Andy replied:
I removed this code (see below comment) but I thought about this more, and I think `else` is safer than `elsif` here since any unexpected `params[:type_of]` that is sent would automatically be assumed to require admin level privileges.

vaidehijoshi commented on app/controllers/response_templates_controller.rb:
```diff
@@ -0,0 +1,74 @@
+  end
+
+  def destroy
+    @response_template = ResponseTemplate.find(params[:id])
```
Any reason to not memoize the `@response_template`?

Zhao-Andy replied:
Just that I didn't think of it! I'm not 100% clear of when to memoize though; should this be memoize in the other actions as well as `destroy`?

vaidehijoshi replied:
Looks like we have `@response_template` defined in `create`, `destroy, and `update`. The `create` method is a little different, but you could check for the `params[:id]` and pull it out into something like this:

```
def response_template
  @response_template ||= if params[:id].present?
    ResponseTemplate.find(params[:id])
  else
    ResponseTemplate.new(permitted_attributes(ResponseTemplate))
end
```

Not sure if this _exact_ code is actually right, but that's what I could imagine. Basically, we only want to memoize this if we know that it's not going to change, and it seems like, we're either `find`ing or creating a `new` response template, so this should be okay I think 🤔

Zhao-Andy replied:
Ohh interesting, thanks for that! That makes a lot more sense. Will add it in!

rhymes commented:
Haven't reviewed this at all but I'm leaving this comment to say: OMG I can't wait :D 

Thanks @Zhao-Andy!

vaidehijoshi commented on app/views/users/_response_templates.html.erb:
```diff
@@ -0,0 +1,35 @@
+<h3>Personal Response Templates</h3>
+<p>Response templates are snippets that you can re-use for quick and accurate comments.</p>
+
+<% if @response_templates.size.positive? %>
```
Ah yeah, this is a good call.

Zhao-Andy commented on app/controllers/users_controller.rb:
```diff
@@ -241,6 +241,8 @@ def handle_settings_tab
       handle_pro_membership_tab
     when "account"
       handle_account_tab
+    when "response-templates"
```
Yeah, this is definitely a place that's growing in technical debt.

Zhao-Andy commented:
I removed the index action altogether since it's technically not used in any of the functionality I'm adding (don't need to refactor if there's no code right?!) 🙃 But this should make things a bit simpler. Also added specs for the CRUD actions and added some better sad path handling, albeit again with a weird implementation...

Zhao-Andy commented:
Yep, the link is available but hidden for now because there's no way to actually use the response templates.

And yeah the index route will come back later. It's meant as a JSON route for async JS to hit.

Zhao-Andy commented on app/views/users/_response_templates.html.erb:
```diff
@@ -0,0 +1,39 @@
+  <h3>Your templates</h3>
+  <div class="github-repos-container">
+    <div class="github-repos">
+    <% @response_templates.each do |response_template| %>
```
Great explanation, thanks!

jacobherrington replied:
TIL, thanks @rhymes :smile:
